% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/data.R
\docType{data}
\name{BST1_LD_matrix}
\alias{BST1_LD_matrix}
\title{LD with the lead SNP: BST1 locus}
\format{
data.table
\describe{
  \item{SNP}{SNP RSID}
  \item{CHR}{Chromosome}
  \item{POS}{Genomic position (in basepairs)}
  \item{...}{Optional: extra columns}
}

\href{https://www.ukbiobank.ac.uk}{UK Biobank}
\href{https://www.biorxiv.org/content/10.1101/807792v3}{Nalls 2019}

matrix
}
\source{
\code{
data("BST1")
finemap_DT <- BST1
# Only including a small subset of the full
# LD matrix for storage purposes.
lead_snp <- subset(finemap_DT, leadSNP)$SNP
snp_list <- finemap_DT[which(finemap_DT$SNP == lead_snp) - 100:which(finemap_DT$SNP == lead_snp) + 100, ]$SNP
BST1_LD_matrix <- readRDS("../Fine_Mapping/Data/GWAS/Nalls23andMe_2019/BST1/plink/UKB_LD.RDS")
BST1_LD_matrix <- BST1_LD_matrix[snp_list, snp_list]
usethis::use_data(BST1_LD_matrix, overwrite = T)
}
}
\usage{
data("BST1_LD_matrix")
}
\description{
Precomputed LD within the \emph{BST1} locus
 (defined in \code{\link[=BST1]{BST1}}.
LD derived British, European-decent subpopulation in the UK Biobank.
Only includes a subset of all the SNPs for storage purposes
(including the lead GWAS/QTL SNP).
}
\details{
Data originally comes from \href{https://www.ukbiobank.ac.uk}{UK Biobank}.
LD was pre-computed and stored by the Alkes Price lab
(see \href{https://www.biorxiv.org/content/10.1101/807792v3}{here}).
}
\keyword{datasets}
