<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting Started â€¢ echoLD</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started">
<meta property="og:description" content="echoLD">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">echoLD</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/echoLD.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RajLabMSSM/echoLD/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting Started</h1>
                        <h4 data-toc-skip class="author"><h4>
Authors: <i>Brian Schilder, Jack Humphrey, Towfique Raj</i>
</h4></h4>
            
            <h4 data-toc-skip class="date"><h4>
Vignette updated: <i>Mar-01-2022</i>
</h4></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/RajLabMSSM/echoLD/blob/HEAD/vignettes/echoLD.Rmd" class="external-link"><code>vignettes/echoLD.Rmd</code></a></small>
      <div class="hidden name"><code>echoLD.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RajLabMSSM/echoLD" class="external-link">echoLD</a></span><span class="op">)</span>

<span class="va">BST1</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/echodata/man/BST1.html" class="external-link">BST1</a></span>
<span class="va">locus_dir</span> <span class="op">&lt;-</span> <span class="fu">echodata</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/echodata/man/locus_dir.html" class="external-link">locus_dir</a></span>

<span class="va">BST1</span> <span class="op">&lt;-</span> <span class="va">BST1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">100</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">locus_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="va">locus_dir</span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="genomes-phase-1-or-3">1000 Genomes: Phase 1 or 3<a class="anchor" aria-label="anchor" href="#genomes-phase-1-or-3"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">LD_1kgp3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_or_create.html">load_or_create</a></span><span class="op">(</span>locus_dir <span class="op">=</span> <span class="va">locus_dir</span>,
                           dat <span class="op">=</span> <span class="va">BST1</span>, 
                           LD_reference <span class="op">=</span> <span class="st">"1KGphase3"</span><span class="op">)</span> <span class="co"># 1KGphase1</span></code></pre></div>
<pre><code><span class="co">## echoLD:: Using 1000Genomes as LD reference panel.</span></code></pre>
<pre><code><span class="co">## LD Reference Panel = 1KGphase3</span></code></pre>
<pre><code><span class="co">## + echoLD:: Querying 1KG remote server.</span></code></pre>
<pre><code><span class="co">## Querying VCF tabix file.</span></code></pre>
<pre><code><span class="co">## LD:: Querying VCF file.</span></code></pre>
<pre><code><span class="co">## Removing non-overlapping SNPs.</span></code></pre>
<pre><code><span class="co">## Time difference of 31.7 secs</span></code></pre>
<pre><code><span class="co">## Saving VCF subset ==&gt; /tmp/RtmpGXtyHi/results/GWAS/Nalls23andMe_2019/BST1/LD/BST1.1kgphase3.vcf.bgz</span></code></pre>
<pre><code><span class="co">## Warning in .Call(.make_vcf_geno, filename, fixed, names(geno), as.list(geno), :</span>
<span class="co">## converting NULL pointer to R NULL</span></code></pre>
<pre><code><span class="co">## Returning VCF with 100 SNPs across 2,504 samples.</span></code></pre>
<pre><code><span class="co">## echoLD::snpStats:: `MAF` column already present.</span></code></pre>
<pre><code><span class="co">## LD:snpStats:: Computing pairwise LD between 2,504 SNPs across 100 individuals (stats = R).</span></code></pre>
<pre><code><span class="co">## Time difference of 0.1 secs</span></code></pre>
<pre><code><span class="co">## Converting obj to sparseMatrix.</span></code></pre>
<pre><code><span class="co">## + echoLD:: Saving LD matrix ==&gt; /tmp/RtmpGXtyHi/results/GWAS/Nalls23andMe_2019/BST1/LD/BST1.1KGphase3_LD.RDS</span></code></pre>
<pre><code><span class="co">## 100 x 100 LD_matrix (sparse)</span></code></pre>
<div class="section level3">
<h3 id="plot">Plot<a class="anchor" aria-label="anchor" href="#plot"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_LD.html">plot_LD</a></span><span class="op">(</span>LD_matrix <span class="op">=</span> <span class="va">LD_1kgp3</span><span class="op">$</span><span class="va">LD</span>, 
        dat <span class="op">=</span> <span class="va">LD_1kgp3</span><span class="op">$</span><span class="va">DT</span>,
        span <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="echoLD_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="uk-biobank">UK Biobank<a class="anchor" aria-label="anchor" href="#uk-biobank"></a>
</h2>
<p><em>WARNING</em>: Takes substantially longer than 1000 Genomes
methods.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">LD_ukb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_or_create.html">load_or_create</a></span><span class="op">(</span>locus_dir <span class="op">=</span> <span class="va">locus_dir</span>,
                         dat <span class="op">=</span> <span class="va">BST1</span>, 
                         LD_reference <span class="op">=</span> <span class="st">"UKB"</span>, 
                         download_method <span class="op">=</span> <span class="st">"axel"</span>, 
                         nThread <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="custom-vcf">Custom VCF<a class="anchor" aria-label="anchor" href="#custom-vcf"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">LD_reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"BST1.1KGphase3.vcf.bgz"</span>, 
                            package <span class="op">=</span> <span class="st">"echoLD"</span><span class="op">)</span>
<span class="va">LD_custom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_or_create.html">load_or_create</a></span><span class="op">(</span>locus_dir <span class="op">=</span> <span class="va">locus_dir</span>,
                            dat <span class="op">=</span> <span class="va">BST1</span>, 
                            LD_reference <span class="op">=</span> <span class="va">LD_reference</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R Under development (unstable) (2022-02-25 r81808)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.3 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">## [1] snpStats_1.45.0  Matrix_1.4-0     survival_3.2-13  echoLD_0.99.1   </span>
<span class="co">## [5] BiocStyle_2.23.1</span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] rjson_0.2.21                ellipsis_0.3.2             </span>
<span class="co">##   [3] rprojroot_2.0.2             XVector_0.35.0             </span>
<span class="co">##   [5] GenomicRanges_1.47.6        fs_1.5.2                   </span>
<span class="co">##   [7] DT_0.21                     gh_1.3.0                   </span>
<span class="co">##   [9] bit64_4.0.5                 lubridate_1.8.0            </span>
<span class="co">##  [11] AnnotationDbi_1.57.1        fansi_1.0.2                </span>
<span class="co">##  [13] xml2_1.3.3                  splines_4.2.0              </span>
<span class="co">##  [15] R.methodsS3_1.8.1           cachem_1.0.6               </span>
<span class="co">##  [17] knitr_1.37                  jsonlite_1.8.0             </span>
<span class="co">##  [19] Rsamtools_2.11.0            dbplyr_2.1.1               </span>
<span class="co">##  [21] png_0.1-7                   R.oo_1.24.0                </span>
<span class="co">##  [23] echodata_0.99.6             BiocManager_1.30.16        </span>
<span class="co">##  [25] compiler_4.2.0              httr_1.4.2                 </span>
<span class="co">##  [27] assertthat_0.2.1            fastmap_1.1.0              </span>
<span class="co">##  [29] cli_3.2.0                   htmltools_0.5.2            </span>
<span class="co">##  [31] prettyunits_1.1.1           tools_4.2.0                </span>
<span class="co">##  [33] glue_1.6.2                  GenomeInfoDbData_1.2.7     </span>
<span class="co">##  [35] dplyr_1.0.8                 rappdirs_0.3.3             </span>
<span class="co">##  [37] Rcpp_1.0.8                  Biobase_2.55.0             </span>
<span class="co">##  [39] jquerylib_0.1.4             pkgdown_2.0.2.9000         </span>
<span class="co">##  [41] vctrs_0.3.8                 Biostrings_2.63.1          </span>
<span class="co">##  [43] echotabix_0.99.2            rtracklayer_1.55.3         </span>
<span class="co">##  [45] xfun_0.29                   stringr_1.4.0              </span>
<span class="co">##  [47] openxlsx_4.2.5              lifecycle_1.0.1            </span>
<span class="co">##  [49] restfulr_0.0.13             XML_3.99-0.9               </span>
<span class="co">##  [51] zlibbioc_1.41.0             BSgenome_1.63.4            </span>
<span class="co">##  [53] VariantAnnotation_1.41.3    ragg_1.2.2                 </span>
<span class="co">##  [55] clisymbols_1.2.0            hms_1.1.1                  </span>
<span class="co">##  [57] MatrixGenerics_1.7.0        parallel_4.2.0             </span>
<span class="co">##  [59] SummarizedExperiment_1.25.3 yaml_2.3.5                 </span>
<span class="co">##  [61] curl_4.3.2                  memoise_2.0.1              </span>
<span class="co">##  [63] reticulate_1.24             sass_0.4.0                 </span>
<span class="co">##  [65] biomaRt_2.51.3              stringi_1.7.6              </span>
<span class="co">##  [67] RSQLite_2.2.10              highr_0.9                  </span>
<span class="co">##  [69] S4Vectors_0.33.10           BiocIO_1.5.0               </span>
<span class="co">##  [71] desc_1.4.0                  GenomicFeatures_1.47.12    </span>
<span class="co">##  [73] BiocGenerics_0.41.2         filelock_1.0.2             </span>
<span class="co">##  [75] zip_2.2.0                   BiocParallel_1.29.15       </span>
<span class="co">##  [77] GenomeInfoDb_1.31.4         rlang_1.0.1                </span>
<span class="co">##  [79] pkgconfig_2.0.3             systemfonts_1.0.4          </span>
<span class="co">##  [81] matrixStats_0.61.0          bitops_1.0-7               </span>
<span class="co">##  [83] evaluate_0.15               lattice_0.20-45            </span>
<span class="co">##  [85] purrr_0.3.4                 GenomicAlignments_1.31.2   </span>
<span class="co">##  [87] htmlwidgets_1.5.4           bit_4.0.4                  </span>
<span class="co">##  [89] tidyselect_1.1.2            magrittr_2.0.2             </span>
<span class="co">##  [91] bookdown_0.24               R6_2.5.1                   </span>
<span class="co">##  [93] IRanges_2.29.1              generics_0.1.2             </span>
<span class="co">##  [95] piggyback_0.1.1             DelayedArray_0.21.2        </span>
<span class="co">##  [97] DBI_1.1.2                   pillar_1.7.0               </span>
<span class="co">##  [99] KEGGREST_1.35.0             RCurl_1.98-1.6             </span>
<span class="co">## [101] tibble_3.1.6                seqminer_8.4               </span>
<span class="co">## [103] crayon_1.5.0                utf8_1.2.2                 </span>
<span class="co">## [105] BiocFileCache_2.3.4         rmarkdown_2.11             </span>
<span class="co">## [107] progress_1.2.2              grid_4.2.0                 </span>
<span class="co">## [109] data.table_1.14.2           blob_1.2.2                 </span>
<span class="co">## [111] digest_0.6.29               tidyr_1.2.0                </span>
<span class="co">## [113] R.utils_2.11.0              textshaping_0.3.6          </span>
<span class="co">## [115] stats4_4.2.0                bslib_0.3.1</span></code></pre>
</details><p><br></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
